File: drivers/verify_kat.c
/*
 * Standalone C verifier for HAWK weak-key NIST PQC `.rsp` KAT files.
 *
 * Calls `hawk_verify_finish` from the unmodified vendored HAWK Round 2 reference
 * implementation in `third_party/hawk-sign-submission/.../Reference_Implementation/`.
 *
 * Usage:
 *     verify_kat path/to/file1.rsp [path/to/file2.rsp ...]
 *
 * The parameter set (logn) is auto-detected from the public-key length.  Exits 0
 * iff every record verifies as `accept`.  Per the artifact's claims, every record
 * in every shipped `.rsp` should accept; any reject is a regression.
 *
 * Build: see drivers/Makefile.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include "hawk.h"
#include "sha3.h"

/* HAWK_PUBKEY_SIZE expands to a constant-fold, so logn must be known at compile
 * time only for the macro form.  We store the three concrete sizes here so we
 * can detect the parameter set from a public key's byte length. */
struct param_entry {
    unsigned logn;
    size_t lenpub;
    size_t lensig;
    const char *name;
};

static const struct param_entry PARAM_TABLE[] = {
    /* Hawk-256:  lenpub = 450 + (256/8 * (5+9))/8 + ... = 450        (per HAWK Round 2 spec) */
    { 8,  450u + ((1u << 8) * (5u + 9u)) / 16u, 249u + (((1u << 8) * (5u + 9u)) / 16u),       "Hawk-256"  },
    { 9,  450u + ((1u << 9) * (5u + 9u)) / 16u, 249u + (((1u << 9) * (5u + 9u)) / 16u),       "Hawk-512"  },
    { 10, 450u + ((1u << 10) * (6u + 10u)) / 16u, 249u + (((1u << 10) * (6u + 10u)) / 16u),   "Hawk-1024" },
};

static int detect_logn(size_t pk_len, unsigned *logn_out, const char **name_out, size_t *lensig_out)
{
    /* Authoritative sizes via the macro, since the formula above is a sanity
     * cross-check of the spec table. */
    static const unsigned char_logn[] = { 8, 9, 10 };
    for (size_t i = 0; i < sizeof char_logn / sizeof char_logn[0]; i++) {
        unsigned ln = char_logn[i];
        size_t expected = 0;
        const char *name = NULL;
        switch (ln) {
            case 8:  expected = HAWK_PUBKEY_SIZE(8);  name = "Hawk-256";  break;
            case 9:  expected = HAWK_PUBKEY_SIZE(9);  name = "Hawk-512";  break;
            case 10: expected = HAWK_PUBKEY_SIZE(10); name = "Hawk-1024"; break;
        }
        if (expected == pk_len) {
            *logn_out = ln;
            *name_out = name;
            switch (ln) {
                case 8:  *lensig_out = HAWK_SIG_SIZE(8);  break;
                case 9:  *lensig_out = HAWK_SIG_SIZE(9);  break;
                case 10: *lensig_out = HAWK_SIG_SIZE(10); break;
            }
            return 0;
        }
    }
    (void)PARAM_TABLE;  /* silence unused if we ever delete the cross-check */
    return -1;
}

static int hex_nibble(int c)
{
    if (c >= '0' && c <= '9') return c - '0';
    if (c >= 'a' && c <= 'f') return c - 'a' + 10;
    if (c >= 'A' && c <= 'F') return c - 'A' + 10;
    return -1;
}

/* Parse a hex string, allocating `*out` on success.  Returns 0 on success. */
static int parse_hex(const char *s, unsigned char **out, size_t *out_len)
{
    size_t n = strlen(s);
    while (n > 0 && isspace((unsigned char)s[n - 1])) n--;
    if (n % 2 != 0) return -1;
    unsigned char *buf = (unsigned char *)malloc(n / 2 + 1);
    if (!buf) return -1;
    for (size_t i = 0; i < n / 2; i++) {
        int hi = hex_nibble(s[2 * i]);
        int lo = hex_nibble(s[2 * i + 1]);
        if (hi < 0 || lo < 0) { free(buf); return -1; }
        buf[i] = (unsigned char)((hi << 4) | lo);
    }
    *out = buf;
    *out_len = n / 2;
    return 0;
}

struct record {
    long count;
    unsigned char *msg;     size_t mlen;
    unsigned char *pk;      size_t pklen;
    unsigned char *sm;      size_t smlen;
};

static void free_record(struct record *r)
{
    free(r->msg); r->msg = NULL;
    free(r->pk);  r->pk  = NULL;
    free(r->sm);  r->sm  = NULL;
    r->count = -1;
}

/* Strip leading whitespace. */
static const char *lskip(const char *s)
{
    while (*s && isspace((unsigned char)*s)) s++;
    return s;
}

/* Read the value half of a `key = value` line, with the leading "key = " already
 * skipped.  Returns a malloc'd null-terminated trimmed copy, or NULL on OOM. */
static char *dup_trim(const char *s)
{
    while (*s && isspace((unsigned char)*s)) s++;
    size_t n = strlen(s);
    while (n > 0 && isspace((unsigned char)s[n - 1])) n--;
    char *out = (char *)malloc(n + 1);
    if (!out) return NULL;
    memcpy(out, s, n);
    out[n] = 0;
    return out;
}

/* Process a complete record: invoke verifier and report. */
static int verify_record(const char *file, struct record *r,
    long *accepts_out, long *total_out)
{
    if (!r->pk || !r->sm) {
        fprintf(stderr, "record %ld in %s missing pk or sm\n", r->count, file);
        return -1;
    }

    unsigned logn = 0;
    const char *name = NULL;
    size_t lensig = 0;
    if (detect_logn(r->pklen, &logn, &name, &lensig) != 0) {
        fprintf(stderr, "record %ld in %s: unrecognised pk length %zu\n",
                r->count, file, r->pklen);
        return -1;
    }
    if (r->smlen < lensig) {
        fprintf(stderr, "record %ld in %s: smlen %zu < lensig %zu\n",
                r->count, file, r->smlen, lensig);
        return -1;
    }
    size_t mlen_eff = r->smlen - lensig;
    if (r->mlen != mlen_eff || (mlen_eff > 0 && memcmp(r->sm, r->msg, mlen_eff) != 0)) {
        fprintf(stderr, "record %ld in %s: sm prefix does not match msg\n", r->count, file);
        return -1;
    }
    const unsigned char *sig = r->sm + mlen_eff;

    size_t tmp_size;
    switch (logn) {
        case 8:  tmp_size = HAWK_TMPSIZE_VERIFY(8);  break;
        case 9:  tmp_size = HAWK_TMPSIZE_VERIFY(9);  break;
        case 10: tmp_size = HAWK_TMPSIZE_VERIFY(10); break;
        default: return -1;
    }
    unsigned char *tmp = (unsigned char *)malloc(tmp_size);
    if (!tmp) { fprintf(stderr, "OOM\n"); return -1; }

    shake_context sc;
    hawk_verify_start(&sc);
    shake_inject(&sc, r->msg, r->mlen);
    int ok = hawk_verify_finish(logn,
        sig, lensig,
        &sc,
        r->pk, r->pklen,
        tmp, tmp_size);
    free(tmp);

    (*total_out)++;
    if (ok) {
        (*accepts_out)++;
    } else {
        fprintf(stderr, "REJECT  %s  count=%ld  alg=%s\n", file, r->count, name);
    }
    return 0;
}

static int verify_file(const char *path, long *accepts_out, long *total_out, const char **alg_out)
{
    FILE *f = fopen(path, "r");
    if (!f) {
        fprintf(stderr, "cannot open %s\n", path);
        return -1;
    }

    char line[16384];
    struct record cur;
    memset(&cur, 0, sizeof cur);
    cur.count = -1;
    *alg_out = NULL;

    long accepts = 0, total = 0;

    while (fgets(line, sizeof line, f)) {
        const char *s = lskip(line);
        if (*s == '\0' || *s == '\n') {
            if (cur.count >= 0 && cur.pk && cur.sm) {
                if (verify_record(path, &cur, &accepts, &total) != 0) {
                    fclose(f);
                    free_record(&cur);
                    return -1;
                }
            }
            free_record(&cur);
            continue;
        }
        if (*s == '#') {
            if (!*alg_out) {
                /* Try to capture the algorithm name from the first non-empty comment. */
                const char *p = s + 1;
                while (*p == ' ' || *p == '#') p++;
                if (strncmp(p, "Hawk-", 5) == 0) {
                    static char buf[64];
                    size_t n = strcspn(p, " \r\n\t");
                    if (n >= sizeof buf) n = sizeof buf - 1;
                    memcpy(buf, p, n);
                    buf[n] = 0;
                    *alg_out = buf;
                }
            }
            continue;
        }

        const char *eq = strchr(s, '=');
        if (!eq) continue;
        size_t klen = (size_t)(eq - s);
        while (klen > 0 && isspace((unsigned char)s[klen - 1])) klen--;
        char key[64];
        if (klen >= sizeof key) continue;
        memcpy(key, s, klen);
        key[klen] = 0;
        const char *valstart = eq + 1;

        if (strcmp(key, "count") == 0) {
            if (cur.pk && cur.sm) {
                if (verify_record(path, &cur, &accepts, &total) != 0) {
                    fclose(f);
                    free_record(&cur);
                    return -1;
                }
            }
            free_record(&cur);
            cur.count = strtol(valstart, NULL, 10);
        } else if (strcmp(key, "mlen") == 0) {
            cur.mlen = (size_t)strtol(valstart, NULL, 10);
        } else if (strcmp(key, "msg") == 0) {
            char *v = dup_trim(valstart);
            if (!v) { fclose(f); return -1; }
            unsigned char *buf; size_t blen;
            if (parse_hex(v, &buf, &blen) != 0) { free(v); fclose(f); return -1; }
            free(v);
            free(cur.msg);
            cur.msg = buf;
            cur.mlen = blen;
        } else if (strcmp(key, "pk") == 0) {
            char *v = dup_trim(valstart);
            if (!v) { fclose(f); return -1; }
            unsigned char *buf; size_t blen;
            if (parse_hex(v, &buf, &blen) != 0) { free(v); fclose(f); return -1; }
            free(v);
            free(cur.pk);
            cur.pk = buf;
            cur.pklen = blen;
        } else if (strcmp(key, "smlen") == 0) {
            cur.smlen = (size_t)strtol(valstart, NULL, 10);
        } else if (strcmp(key, "sm") == 0) {
            char *v = dup_trim(valstart);
            if (!v) { fclose(f); return -1; }
            unsigned char *buf; size_t blen;
            if (parse_hex(v, &buf, &blen) != 0) { free(v); fclose(f); return -1; }
            free(v);
            free(cur.sm);
            cur.sm = buf;
            cur.smlen = blen;
        }
        /* seed and sk are intentionally ignored. */
    }

    if (cur.count >= 0 && cur.pk && cur.sm) {
        if (verify_record(path, &cur, &accepts, &total) != 0) {
            fclose(f);
            free_record(&cur);
            return -1;
        }
    }
    free_record(&cur);
    fclose(f);

    *accepts_out = accepts;
    *total_out = total;
    return 0;
}

int main(int argc, char **argv)
{
    if (argc < 2) {
        fprintf(stderr, "usage: %s file.rsp [...]\n", argv[0]);
        return 2;
    }

    int failures = 0;
    for (int i = 1; i < argc; i++) {
        long accepts = 0, total = 0;
        const char *alg = NULL;
        if (verify_file(argv[i], &accepts, &total, &alg) != 0) {
            failures++;
            continue;
        }
        const char *base = strrchr(argv[i], '/');
        base = base ? base + 1 : argv[i];
        const char *status = (accepts == total) ? "PASS" : "FAIL";
        printf("[%s] c-ref       %-40s  alg=%-32s  %ld/%ld accept\n",
               status, base, alg ? alg : "<unknown>", accepts, total);
        if (accepts != total) failures++;
    }
    return failures == 0 ? 0 : 1;
}

File: drivers/verify_kat.py
"""Verify HAWK weak-key NIST PQC `.rsp` KAT files using the official Python reference.

Calls `verify.hawkverify` from the unmodified HAWK Round 2 submission's `Extra/hawk-py/`.
This is one of the two independent verifiers in this artifact (the other is the C reference
in `drivers/verify_kat.c`).

Usage:

    python3 drivers/verify_kat.py vectors/kat/mbs_hawk512.rsp [more.rsp ...]

Exits 0 iff every record verifies as `accept`.  Per the artifact's claims, every record
in every shipped `.rsp` should accept; a single reject is a regression.
"""

from __future__ import annotations

import argparse
import re
import sys
from dataclasses import dataclass
from pathlib import Path

import numpy as np

ROOT = Path(__file__).resolve().parents[1]
HAWK_PY = ROOT / "third_party" / "hawk-sign-submission" / "hawk-submission" / "Extra" / "hawk-py"
sys.path.insert(0, str(HAWK_PY))

from params import PARAMS  # noqa: E402  official parameter table
from verify import hawkverify  # noqa: E402  official Python verifier (unmodified)


PARAM_LEN_TO_LOGN = {
    PARAMS(8, "lenpub"): 8,
    PARAMS(9, "lenpub"): 9,
    PARAMS(10, "lenpub"): 10,
}


@dataclass
class Record:
    count: int
    msg: bytes
    pk: bytes
    sm: bytes


def parse_rsp(path: Path) -> tuple[str, list[Record]]:
    """Parse a NIST PQC `.rsp` file into (algorithm_name, records).

    The parser tolerates `# ...` comment lines anywhere; key=value lines are
    grouped into records by `count = N`.
    """
    alg_name = ""
    records: list[Record] = []
    current: dict[str, str] = {}

    with path.open() as fh:
        for raw in fh:
            line = raw.strip()
            if not line:
                if current:
                    records.append(_record_from_dict(current))
                    current = {}
                continue
            if line.startswith("#"):
                if not alg_name and len(line) > 1:
                    alg_name = line.lstrip("# ").strip()
                continue
            m = re.match(r"^([a-zA-Z_]+)\s*=\s*(.*)$", line)
            if not m:
                continue
            key, val = m.group(1), m.group(2)
            if key == "count" and current:
                records.append(_record_from_dict(current))
                current = {}
            current[key] = val

    if current:
        records.append(_record_from_dict(current))

    return alg_name, records


def _record_from_dict(d: dict[str, str]) -> Record:
    count = int(d["count"])
    msg = bytes.fromhex(d["msg"]) if d.get("msg") else b""
    pk = bytes.fromhex(d["pk"])
    sm = bytes.fromhex(d["sm"])
    return Record(count=count, msg=msg, pk=pk, sm=sm)


def verify_one(record: Record) -> tuple[bool, int]:
    """Return (accept, logn).  Raises on unrecognised public key length."""
    logn = PARAM_LEN_TO_LOGN.get(len(record.pk))
    if logn is None:
        raise ValueError(f"public key length {len(record.pk)} does not match any HAWK parameter set")
    sig_len = PARAMS(logn, "lensig")
    if len(record.sm) < sig_len:
        raise ValueError(f"sm length {len(record.sm)} < lensig {sig_len}")
    msg = record.sm[: len(record.sm) - sig_len]
    sig = record.sm[len(record.sm) - sig_len:]
    if msg != record.msg:
        raise ValueError("sm prefix does not match msg")

    pk_arr = np.frombuffer(record.pk, dtype=np.uint8)
    msg_arr = np.frombuffer(msg, dtype=np.uint8)
    sig_arr = np.frombuffer(sig, dtype=np.uint8)
    return bool(hawkverify(logn, pk_arr, msg_arr, sig_arr)), logn


def verify_file(path: Path, max_records: int | None = None) -> tuple[int, int, str, int]:
    """Return (accepts, checked, alg_name, total_in_file) for one `.rsp` file."""
    alg_name, records = parse_rsp(path)
    total_in_file = len(records)
    if max_records is not None:
        records = records[:max_records]
    accepts = 0
    for rec in records:
        ok, _logn = verify_one(rec)
        if ok:
            accepts += 1
        else:
            print(f"REJECT  {path.name}  count={rec.count}", file=sys.stderr)
    return accepts, len(records), alg_name, total_in_file


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("paths", nargs="+", help=".rsp file(s) to verify")
    parser.add_argument("--max-records", type=int, default=None,
                        help="verify at most N records per file (default: all)")
    args = parser.parse_args()

    failures = 0
    for p in args.paths:
        path = Path(p)
        accepts, checked, alg, total = verify_file(path, args.max_records)
        status = "PASS" if accepts == checked else "FAIL"
        suffix = f"({checked}/{total} sampled)" if checked < total else ""
        print(f"[{status}] python-ref  {path.name:40s}  alg={alg:32s}  {accepts}/{checked} accept {suffix}")
        if accepts != checked:
            failures += 1
    return 0 if failures == 0 else 1


if __name__ == "__main__":
    raise SystemExit(main())
